docs/en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_ReactJS_Redux_and_GraphQL.md

```diff
@@ -0,0 +1,666 @@
+
+Much like customisation is done in the backend through [Injector](../../extending/injector), the client-side framework has its own implementation of dependency injection known as `Injector`. Using Injector, you can register new services, and transform existing services.
+
+Injector is broken up into three sub-APIs:
+* `Injector.react` for React UI components
+* `Injector.reducer` for Redux state management
+* `Injector.form` for forms rendered via `FormSchema`.
+
+The frontend Injector works a bit differently than its backend counterpart. Instead of _overriding_ a service with your own implementation, you _enhance_ an existing service with your own concerns. This pattern is known as [middleware](https://en.wikipedia.org/wiki/Middleware).
+
+Middleware works a lot like a decorator. It doesn't alter the original API of the service,
+but it can augment it with new features and concerns. This has the inherent advantage of allowing all thidparty code to have an influence over the behaviour, state, and UI of a component.
+
+## A simple middleware example
+
+Let's say you have an application that features error logging. By default, the error logging service simply outputs to `console.error`. But you want to customise it to send errors to a thirdparty service. For this, you could use middleware to augment the default functionality of the logger.
+
+_LoggingService.js_
+```js
+(error) => console.error(error);
+```
+
+Now, let's add some middleware to that service. The signature of middleware is: 
+```js
+(next) => (args) => next(args)
+```
+Where `next()` is the next customisation in the "chain" of middleware. Before invoking the next implementation, you can add whatever customisations you need. Here's how we would use middleware to enhance `LoggingService`.
+
+```js
+(LoggingService) => (error) => {
+	if (error.type === LoggingService.CRITICAL) {
+		sendToThirdpartyLoggingService(error.message);
+	}
+	return LoggingService(error);
+}
+```
+
+We haven't overriden any functionality. `LoggingService(error)` will still invoke `console.error`, once all the middleware has run. But what if we did want to kill the original functionality?
+
+```js
+(LoggingService) => (error) => {
+	// Critical errors go to a thirdparty service
+	if (error.type === LoggingService.CRITICAL) {
+		sendToThirdpartyLoggingService(error.message);
+	}
+	// Other errors get logged, but not to our thirdparty
+	else if (error.type === LoggingService.ERROR) {
+		LoggingService(error);
+	} 
+	// Minor errors are ignored
+	else {
+		// Do nothing!
+	}
+}
+```
+
+
+## Registering new services to the Injector
+
+If you've created a module using React, it's a good idea to afford other developers an 
+API to enhance those components, forms, and state. To do that, simply register them with `Injector`.
+
+__my-public-module/js/main.js__
+```js
+import Injector from 'lib/Injector';
+
+Injector.react.register('MyComponent', MyComponent);
+Injector.reducer.register('myCustom', MyReducer);
```

Comment:
This should also mention how the injector services can be used then, otherwise it'll be hard to understand why this is needed in the first place. Probably through the `MyComponent` example, since that's the easiest to understand for anybody with React context.